### PR TITLE
error in bake() if variable is missing in new_data

### LINKED ISSUE #491: When one column missing in new_data that is used in recipe, bake() fails uniformatively
Can we fail more informatively here?
``` r
library(recipes)
#> Loading required package: dplyr
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
#> 
#> Attaching package: 'recipes'
#> The following object is masked from 'package:stats':
#> 
#>     step
library(datasets)
rec <- recipe(event ~ ., data = attenu) %>%
  step_integer(station)
prepped <- prep(rec)
bake(prepped, attenu %>% dplyr::select(-.data$station))
#> Error: Can't join on 'value' x 'value' because of incompatible types (character / NULL)
```

<sup>Created on 2020-04-07 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

<details>

<summary>Session info</summary>

``` r
devtools::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 3.6.3 (2020-02-29)
#>  os       Ubuntu 18.04.4 LTS          
#>  system   x86_64, linux-gnu           
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_US.UTF-8                 
#>  ctype    en_US.UTF-8                 
#>  tz       Europe/Zurich               
#>  date     2020-04-07                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version    date       lib source        
#>  assertthat    0.2.1      2019-03-21 [1] CRAN (R 3.6.1)
#>  backports     1.1.5      2019-10-02 [1] CRAN (R 3.6.1)
#>  callr         3.4.2      2020-02-12 [1] CRAN (R 3.6.3)
#>  class         7.3-16     2020-03-25 [4] CRAN (R 3.6.3)
#>  cli           2.0.2      2020-02-28 [1] CRAN (R 3.6.3)
#>  crayon        1.3.4      2017-09-16 [1] CRAN (R 3.6.1)
#>  desc          1.2.0      2018-05-01 [1] CRAN (R 3.6.1)
#>  devtools      2.2.2      2020-02-17 [1] CRAN (R 3.6.3)
#>  digest        0.6.25     2020-02-23 [1] CRAN (R 3.6.3)
#>  dplyr       * 0.8.5      2020-03-07 [1] CRAN (R 3.6.3)
#>  ellipsis      0.3.0      2019-09-20 [1] CRAN (R 3.6.1)
#>  evaluate      0.14       2019-05-28 [1] CRAN (R 3.6.1)
#>  fansi         0.4.1      2020-01-08 [1] CRAN (R 3.6.1)
#>  fs            1.4.1      2020-04-04 [1] CRAN (R 3.6.3)
#>  generics      0.0.2      2018-11-29 [1] CRAN (R 3.6.1)
#>  glue          1.4.0      2020-04-03 [1] CRAN (R 3.6.3)
#>  gower         0.2.1      2019-05-14 [1] CRAN (R 3.6.2)
#>  highr         0.8        2019-03-20 [1] CRAN (R 3.6.1)
#>  htmltools     0.4.0      2019-10-04 [1] CRAN (R 3.6.1)
#>  ipred         0.9-9      2019-04-28 [1] CRAN (R 3.6.2)
#>  knitr         1.28       2020-02-06 [1] CRAN (R 3.6.3)
#>  lattice       0.20-40    2020-02-19 [4] CRAN (R 3.6.2)
#>  lava          1.6.7      2020-03-05 [1] CRAN (R 3.6.3)
#>  lifecycle     0.2.0      2020-03-06 [1] CRAN (R 3.6.3)
#>  lubridate     1.7.4      2018-04-11 [1] CRAN (R 3.6.1)
#>  magrittr    * 1.5        2014-11-22 [1] CRAN (R 3.6.1)
#>  MASS          7.3-51.5   2019-12-20 [4] CRAN (R 3.6.2)
#>  Matrix        1.2-18     2019-11-27 [4] CRAN (R 3.6.1)
#>  memoise       1.1.0      2017-04-21 [1] CRAN (R 3.6.1)
#>  nnet          7.3-13     2020-02-25 [4] CRAN (R 3.6.3)
#>  pillar        1.4.3      2019-12-20 [1] CRAN (R 3.6.1)
#>  pkgbuild      1.0.6      2019-10-09 [1] CRAN (R 3.6.1)
#>  pkgconfig     2.0.3      2019-09-22 [1] CRAN (R 3.6.1)
#>  pkgload       1.0.2      2018-10-29 [1] CRAN (R 3.6.1)
#>  prettyunits   1.1.1      2020-01-24 [1] CRAN (R 3.6.2)
#>  processx      3.4.2      2020-02-09 [1] CRAN (R 3.6.3)
#>  prodlim       2019.11.13 2019-11-17 [1] CRAN (R 3.6.2)
#>  ps            1.3.2      2020-02-13 [1] CRAN (R 3.6.3)
#>  purrr         0.3.3      2019-10-18 [1] CRAN (R 3.6.1)
#>  R6            2.4.1      2019-11-12 [1] CRAN (R 3.6.1)
#>  Rcpp          1.0.4      2020-03-17 [1] CRAN (R 3.6.3)
#>  recipes     * 0.1.10     2020-03-18 [1] CRAN (R 3.6.3)
#>  remotes       2.1.1      2020-02-15 [1] CRAN (R 3.6.3)
#>  rlang         0.4.5      2020-03-01 [1] CRAN (R 3.6.3)
#>  rmarkdown     2.1        2020-01-20 [1] CRAN (R 3.6.3)
#>  rpart         4.1-15     2019-04-12 [4] CRAN (R 3.6.1)
#>  rprojroot     1.3-2      2018-01-03 [1] CRAN (R 3.6.1)
#>  sessioninfo   1.1.1      2018-11-05 [1] CRAN (R 3.6.1)
#>  stringi       1.4.6      2020-02-17 [1] CRAN (R 3.6.3)
#>  stringr       1.4.0      2019-02-10 [1] CRAN (R 3.6.1)
#>  survival      3.1-11     2020-03-07 [4] CRAN (R 3.6.3)
#>  testthat      2.3.2      2020-03-02 [1] CRAN (R 3.6.3)
#>  tibble        3.0.0      2020-03-30 [1] CRAN (R 3.6.3)
#>  tidyr         1.0.2      2020-01-24 [1] CRAN (R 3.6.3)
#>  tidyselect    1.0.0      2020-01-27 [1] CRAN (R 3.6.3)
#>  timeDate      3043.102   2018-02-21 [1] CRAN (R 3.6.2)
#>  usethis     * 1.5.1      2019-07-04 [1] CRAN (R 3.6.1)
#>  vctrs         0.2.4      2020-03-10 [1] CRAN (R 3.6.3)
#>  withr         2.1.2      2018-03-15 [1] CRAN (R 3.6.3)
#>  xfun          0.12       2020-01-13 [1] CRAN (R 3.6.2)
#>  yaml          2.2.1      2020-02-01 [1] CRAN (R 3.6.3)
#> 
#> [1] /home/walor/R/x86_64-pc-linux-gnu-library/3.6
#> [2] /usr/local/lib/R/site-library
#> [3] /usr/lib/R/site-library
#> [4] /usr/lib/R/library
```

</details>

### PR DESCRIPTION
**Pull request by EmilHvitfeldt**
This PR aims to close #491.

It adds a guard in `bake.recipe()` that checks that all the original predictors are available. Errors out otherwise, consistent with similar errors such as from `check_cols()`.

A couple of existing tests were changed to adhere to this higher level of strictness 

# Example:

``` r
library(recipes)
library(modeldata)

data(ames)
rec <- recipe(Sale_Price ~ ., data = ames) %>%
  step_dummy(MS_SubClass, MS_Zoning)
prepped <- prep(rec)

bake(prepped, ames %>% dplyr::select(-MS_SubClass))
#> Error: The following cols are missing from `new_data`: MS_SubClass
bake(prepped, ames %>% dplyr::select(-MS_SubClass, -MS_Zoning))
#> Error: The following cols are missing from `new_data`: MS_SubClass, MS_Zoning
bake(prepped, ames %>% dplyr::select(Sale_Price))
#> Error: The following cols are missing from `new_data`: MS_SubClass, MS_Zoning...
```

### CODE REVIEW

File: R/misc.R [hunk 1]
```diff
@@ -659,6 +659,24 @@ check_training_set <- function(x, rec, fresh) {
   x
 }
 
+check_newdata_columns <- function(object, new_data) {
```
Comment (juliasilge) [inline R/misc.R:662]: Can we call this function `check_new_data_columns()`?

Comment (EmilHvitfeldt) [reply R/misc.R]: Done!

File: R/misc.R [hunk 2]
```diff
@@ -659,6 +659,24 @@ check_training_set <- function(x, rec, fresh) {
   x
 }
 
+check_new_data_columns <- function(object, new_data) {
```
Comment (topepo) [inline R/misc.R:662]: We should have an argument here for role exclusions. We don't want to error if the outcome columns are missing. 

We'll need some time to think about this in regards to what to do about other roles such as case weights, id variables, performance variables and so on.

File: tests/testthat/test_bake.R [hunk 3]
```diff
@@ -21,3 +21,32 @@ test_that("can use tidyselect ops in bake() and juice() column selection", {
...
+  expect_error(rec2 %>% prep() %>% bake(mtcars_mat[, -(2:6)]), error26)
+
+  mtcars_mat <- as.matrix(mtcars)
+  rec3 <- recipe(x = mtcars_mat)
+
+  expect_error(rec3 %>% prep() %>% bake(mtcars_mat), NA)
+  expect_error(rec3 %>% prep() %>% bake(mtcars_mat[, -1]), error1)
+  expect_error(rec3 %>% prep() %>% bake(mtcars_mat[, -2]), error2)
+  expect_error(rec3 %>% prep() %>% bake(mtcars_mat[, -(2:6)]), error26)
+})
```
Comment (DavisVaughan) [inline tests/testthat/test_bake.R:52]: Doesn't this break case weights? i.e. they are required at `prep()` time but not used at `bake()` time - and im pretty sure that triggers an error?

Comment (EmilHvitfeldt) [reply tests/testthat/test_bake.R:52]: It might, this PR is pre case weights. So I wouldn't be surprised if it doesn't work in its current state

**Pull request merged**